pip/pip-376.md

```diff
@@ -0,0 +1,138 @@
+# PIP-376: Make topic policies service pluggable
+
+# Background knowledge
+
+## Topic policies service and system topics
+
+[PIP-39](https://github.com/apache/pulsar/wiki/PIP-39%3A-Namespace-Change-Events) introduces system topics and the topic level policies. However, the topic policies service (`TopicPoliciesService`) only has one implementation (`SystemTopicBasedTopicPoliciesService`) that depends on the system topics. So the following configs are both required (though they're all enabled by default now):
+
+```properties
+systemTopicEnabled=true
+topicLevelPoliciesEnabled=true
+```
+
+However, if the Pulsar storage is switched to a S3-based solution (by modifying the `managedLedgerStorageClassName` config), using system topics to manage topic policies could have low performance (due to the S3 write and read latency) and higher cost (due to redundant S3 API calls).
```

Comment:
>  if the Pulsar storage is switched to a S3-based solution
is this a custom thing ?
I know that `managedLedgerStorageClassName` is configurable, but it is not properly a place where it is possible to code custom implementations because the API boundaries are not clear (as for Protocol Handlers, you end up in using internal classes)

Reply:
Agreed. I will change the specific case here.